### PR TITLE
[eme] Add tests for persistent-license sessions

### PR DESCRIPTION
**Pull request by mwatson2**

### CODE REVIEW

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            assert_true(event instanceof window.Event );
+            assert_equals(event.type, 'keystatuseschange' );
+
+            var hasKeys = false, pendingKeys = false;
+            _mediaKeySession.keyStatuses.forEach( function( value, keyid ) {
+                assert_any( assert_equals, value, [ 'status-pending', 'usable' ] );
+
+                hasKeys = true;
+                pendingKeys = pendingKeys || ( value === 'status-pending' );
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: nit: empty line

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            });
+
+            if ( !_allKeysUsableEvent && hasKeys && !pendingKeys ) {
+                _allKeysUsableEvent = true;
+                _events.push( 'allkeysusable' );
+                _video.setMediaKeys(_mediaKeys);
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Is there a reason to wait until now to call this?

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            }).catch(function(error) {
+                forceTestFailureFromPromise(test, error);
+            });
+        }
+
+        function onClosed(event) {
+            _events.push( 'closed-promise' );
+
+            setTimeout( test.step_func( function() {
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: nit: empty

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                                        'remove',
+                                        'emptykeyslist',
+                                        'license-release',
+                                        'release-response',
+                                        'closed-promise',
+                                        'updated'
+                                    ],
+                                    "Expected events sequence" );
+
+                _video.src = "";
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: OOC, why is this necessary?

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                _video.setMediaKeys( null ).then( function() { test.done() } );
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: nit: empty

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            } ), 0 );
+        }
+
+        function onTimeupdate(event) {
+            if ( _video.currentTime > ( config.duration || 5 ) && !_timeupdateEvent ) {
+                _timeupdateEvent = true;
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: The meanings of these variable names are non-obvious. Is this the style in WPT? If not, something like receivedTimeupdateEvent, TimeupdateEventReceived, hasReceivedTimeupdateEvent, isWaitingForTimeupdateEvent, etc. would be clearer.

File: encrypted-media/scripts/playback-persistent-license.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            forceTestFailureFromPromise(test, error);
+        }
+
+        function onMessage(event) {
+            assert_equals( event.target, _mediaKeySession );
+            assert_true( event instanceof window.MediaKeyMessageEvent );
+            assert_equals( event.type, 'message');
+
+            if ( !_releaseSequence )
+            {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: nit: Opening brace on its own line is inconsistent. Also line 35.

File: encrypted-media/scripts/playback-persistent-license.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        function onFailure(error) {
+            forceTestFailureFromPromise(test, error);
+        }
+
+        function onMessage(event) {
+            assert_equals( event.target, _mediaKeySession );
+            assert_true( event instanceof window.MediaKeyMessageEvent );
+            assert_equals( event.type, 'message');
+
+            if ( !_releaseSequence )
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: This is different from how this check was made in the previous file. Is there a reason for that?

File: encrypted-media/scripts/playback-persistent-license.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            });
+        }
+
+        function onEncrypted(event) {
+            assert_equals(event.target, _video);
+            assert_true(event instanceof window.MediaEncryptedEvent);
+            assert_equals(event.type, 'encrypted');
+
+            waitForEventAndRunStep('message', _mediaKeySession, onMessage, test);
+            _mediaKeySession.generateRequest(   config.initData ? config.initDataType : event.initDataType,
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: Is there a reason not to use `config.initDataType || event.initDataType` here too?

File: encrypted-media/scripts/playback-persistent-license.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+        function onClosed(event) {
+            _video.src = "";
+            _video.setMediaKeys( null ).then(function(){
+                    test.done();
+            });
+        }
+
+        function onTimeupdate(event) {
+            if ( _video.currentTime > ( config.duration || 5 ) && !_releaseSequence ) {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: Why is 5 the default? That's a long test.

File: encrypted-media/scripts/playback-persistent-license.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
+function runTest(config, testname) {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: Is this test a subset of the events one above? Does it test anything additional?

It would be nice to minimize the duplication.

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 12]
```diff
@@ -0,0 +1,168 @@
+function runTest(config, testname) {
+
+    var testname = config.keysystem + ', successful playback, persistent-license, '
+                                    + /video\/([^;]*)/.exec( config.videoType )[ 1 ]
+                                    + ', set src before setMediaKeys, check events';
+
+    var configuration = {   initDataTypes: [ config.initDataType ],
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Is there a reason for the three spaces after the '{' here, before the '?' on line 33, etc.? This just wastes whitespace, especially for longer lines on smaller displays.

File: encrypted-media/scripts/playback-retrieve-persistent-license.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+            config.messagehandler( config.keysystem, event.messageType, event.message ).then( function( response ) {
+
+                _mediaKeySession.update( response )
+                .catch(onFailure);
+            });
+        }
+
+        function onPlaying(event) {
+
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: nit: empty line here and line 60

File: encrypted-media/scripts/playback-retrieve-persistent-license.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                    assert_equals(assertion.actual, assertion.expected, assertion.message);
+
+                }));
+
+                win.close();
+
+                test.done();
+            }));
+
+            delete config.video;
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: Why?

File: encrypted-media/scripts/playback-retrieve-persistent-license.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        }
+
+        function onClosed(event) {
+
+            _video.src = "";
+            _video.setMediaKeys( null );
+
+            var win = window.open( config.windowscript );
+            window.addEventListener('message', test.step_func(function( event ) {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: What is `event`? Is that name important or could a less ambiguous and non-overloaded name be used?

File: encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                        keysystem:          'org.w3.clearkey',
+                        messagehandler:     messagehandler.bind( contentitem ),
+                        audioPath:          contentitem.audio.path,
+                        videoPath:          contentitem.video.path,
+                        audioType:          contentitem.audio.type,
+                        videoType:          contentitem.video.type,
+                        initDataType:       'keyids',
+                        initData:           getInitData(contentitem,'keyids'),
+                        duration:           2,
+                        windowscript:       'resources/retrieve-persistent-license.html' };
```
Comment (ddorwin) [inline encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html]: Why isn't this in the .js file?

File: encrypted-media/clearkey-mp4-playback-persistent-license-events.html [hunk 17]
```diff
@@ -0,0 +1,52 @@
+<!doctype html>
+<html>
+  <head>
+    <meta charset=utf-8>
+    <meta name="timeout" content="long">
+    <title>Encrypted Media Extensions: Successful Playback, persistent-usage-record session with Clear Key, mp4, event sequence</title>
```
Comment (ddorwin) [inline encrypted-media/clearkey-mp4-playback-persistent-license-events.html]: "persistent-usage-record" should be "persistent-license" three places in this PR.

File: encrypted-media/resources/retrieve-persistent-license.html [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+            return _mediaKeySession.load( event.data.sessionId );
+        }).then(function( success ) {
+            if ( !success ) throw new DOMException( 'Could not load session' );
+            return testmediasource(config);
+        }).then(function(source) {
+            config.video.src = URL.createObjectURL(source);
+            config.video.play();
+        })
+        .catch(onFailure);
+
```
Comment (ddorwin) [inline encrypted-media/resources/retrieve-persistent-license.html]: empty lines

File: encrypted-media/resources/retrieve-persistent-license.html [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+  <body>
+    <div id='log'></div>
+
+    <div id='video'>
+      <video id="videoelement" width="200px"></video>
+    </div>
+
+    <script>
+
+    window.addEventListener( 'message', function( event ) {
```
Comment (ddorwin) [inline encrypted-media/resources/retrieve-persistent-license.html]: It might be worth noting that this is from `postMessage()` and not the same as the `MediaKeySession` event.

File: encrypted-media/resources/retrieve-persistent-license.html [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+
+        navigator.requestMediaKeySystemAccess(config.keysystem, [ configuration ] )
+        .then(function(access) {
+            return access.createMediaKeys();
+        }).then(function(mediaKeys) {
+            config.video.setMediaKeys(mediaKeys);
+            config.video.addEventListener('timeupdate', onTimeupdate, true);
+            consoleWrite(event.data.sessionId);
```
Comment (ddorwin) [inline encrypted-media/resources/retrieve-persistent-license.html]: Remove debugging logging?

File: encrypted-media/resources/retrieve-persistent-license.html [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+            config.messagehandler( config.keysystem, event.messageType, event.message )
+            .then( function( response ) {
+                event.target.update( response ).catch(onFailure);
+            });
+        }
+
+        function onTimeupdate(event) {
+            if ( config.video.currentTime > ( config.duration || 5 ) ) {
+                config.video.pause();
+                window.opener.postMessage(assertions, '*');
```
Comment (ddorwin) [inline encrypted-media/resources/retrieve-persistent-license.html]: It seems we should be more explicit about the test being completed successfully. Maybe there should be an explicit value to avoid any false successes.

File: encrypted-media/resources/retrieve-persistent-license.html [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+            }
```
Comment (ddorwin) [inline encrypted-media/resources/retrieve-persistent-license.html]: We should remove the session to clean up after the (successful) test.

**Comment by ddorwin** [comment]
I only reviewed the core files (resources/*). Overall, it seems fine, though I didn't do a thorough review. I'm not aware of any Clear Key implementations that support `"persistent-license"`, but you can test Widevine on Chrome OS.

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 2 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: No. I was probably trying to make it work on Edge, which seems to require setMediaKeys to be called quite late.

We'll need a more comprehensive test for the timing of setMediaKeys - there seems to be more to it than before / after setting the src.

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 4 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: I'm not sure that it is. I'll remove it.

File: encrypted-media/scripts/playback-persistent-license.js (hunk 8 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license.js]: No. I'll align.

File: encrypted-media/scripts/playback-persistent-license.js (hunk 9 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license.js]: Just to be sure that we always supply both the config variable or both the event variables and never one of each.

File: encrypted-media/scripts/playback-persistent-license.js (hunk 10 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license.js]: Yep, that should probably be cleaned up globally. I added it to #3423.

File: encrypted-media/scripts/playback-persistent-license.js (hunk 11 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license.js]: Yes, it is a strict subset. It tests basic playback for this session type. The duplication is necessary because you can't test for the correct playback event sequence without doing playback.

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 12 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: It just lines everything up with a tab position. Are smaller displays really a concern? This line is 20% of the display width on my laptop.

File: encrypted-media/scripts/playback-retrieve-persistent-license.js (hunk 14 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: They can't be structured cloned and so can't be handed over to the other window.

File: encrypted-media/scripts/playback-retrieve-persistent-license.js (hunk 15 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: I'll change to `messageEvent`.

File: encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html (hunk 16 above)
Comment (mwatson2) [inline encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html]: This file is specific to clearkey. It should be named as such. We're going to need a different one for the drm case, but the JS script in the first window can be common, so we specify which second-window script to use here.

Also, we may want further second-window scripts which test different things for the retrieved-license case, for example that it retains the correct expiration. We can use the same first-window JS for these too, with different second-window content.

File: encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html [hunk 23]
```diff
@@ -0,0 +1,53 @@
+<!doctype html>
+<html>
+  <head>
+    <meta charset=utf-8>
+    <meta name="timeout" content="long">
+    <title>Encrypted Media Extensions: Successful Playback, persistent-usage-record session with Clear Key, mp4, event sequence</title>
```
Comment (ddorwin) [inline encrypted-media/clearkey-mp4-playback-retrieve-persistent-license.html]: persistent-license

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 12 above)
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: It can make side-by-side diffing more difficult, and GitHub doesn't handle long lines well. This was a general comment and not specific to this line. It looks like we're using four-space tabs, which has a similar effect. Do WPT have a preferred style?

Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: I don't see any coding style guidelines in the documentation. We can clean this up under #3423.

File: encrypted-media/scripts/playback-persistent-license.js (hunk 9 above)
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: Shouldn't we change the other instance too then?

Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license.js]: Yes, probably.

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+                _video.setMediaKeys( null ).then( function() { test.done() } );
+            } ), 0 );
+        }
+
+        function onTimeupdate(event) {
+            if ( _video.currentTime > ( config.duration || 5 ) && !_receivedTimeupdateEvent ) {
+                _receivedTimeupdateEvent = true;
+                _video.pause();
+                _video.src = "";
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Global comment for this PR and probably all our files:
`.src = ""` is incorrect as it doesn't do what one might think it does. Use ` video.removeAttribute('src');` instead.
Also, to be sure you've freed the media player, you must follow it with `video.load();`

Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Specific comment: Why do we need to clear the source here?

Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: I believe it's allowed for remove() to fail if the license is "in use", but "in use" is not very well defined. So we need to do whatever is necessary to ensure the license is no longer in use. I thought `video.src=""` would be enough, but I can change it as you suggest.

**Comment by mwatson2** [comment]
These tests pass on Chrome / Firefox / Edge using the clearkey polyfill, which relies only on `temporary` session support from the browser.

I will add DRM versions in a separate PR.

File: encrypted-media/scripts/playback-persistent-license.js [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+            assert_equals(event.type, 'encrypted');
+
+            waitForEventAndRunStep('message', _mediaKeySession, onMessage, test);
+            _mediaKeySession.generateRequest(   config.initData ? config.initDataType : event.initDataType,
+                                                config.initData || event.initData )
+            .catch(onFailure);
+        }
+
+        function onClosed(event) {
+            _video.src = "";
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license.js]: Are these two lines relevant?

File: encrypted-media/scripts/playback-retrieve-persistent-license.js [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+                _video.pause();
+
+                _mediaKeySession.closed.then( test.step_func( onClosed ) );
+
+                _mediaKeySession.close();
+            }
+        }
+
+        function onClosed() {
+            _video.src = "";
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-retrieve-persistent-license.js]: What is the purpose of these two lines?

If the purpose here is to destroy the MediaKeys (CDM instance) before loading another one:
1. Probably add a comment as such.
2. Put everything in a .then(). This will at least wait until the media player is no longer using the CDM.
3. Clear _mediaKeys and _mediaKeySession.
Still, there is no guarantee that the underlying CDM [instance] or session has been destroyed due to garbage collection.

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+  <body>
+    <div id='log'></div>
+
+    <div id='video'>
+      <video id="videoelement" width="200px"></video>
+    </div>
+
+    <script>
+
+    window.addEventListener( 'message', function( event ) {
```
Comment (ddorwin) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: It might be worth noting that this is from postMessage() and not the same as the MediaKeySession event.

(There were no responses to the [comments in this file before it was renamed](https://github.com/w3c/web-platform-tests/pull/3424/files/1703c17631dcce9be010c909f32178bf1564d4ac#diff-7a20c3da279e1bbc305b0a68f50a1b7a), so I've duplicated them here.)

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+            assertions = [ ];
+
+            var _mediaKeySession;
+
+            config.video = document.getElementById('videoelement');
+            config.messagehandler = (new MessageHandler( 'org.w3.clearkey')).messagehandler;
+
+        function onFailure(error) {
+            assertions.push( { actual: false, expected: true, message: error.toString() } );
+            window.opener.postMessage(assertions, '*');
```
Comment (ddorwin) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: It seems we should be more explicit about the test being completed successfully. Maybe there should be an explicit value to avoid any false successes.

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+        }
```
Comment (ddorwin) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: We should remove the session to clean up after the (successful) test.

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+
+        navigator.requestMediaKeySystemAccess(config.keysystem, [ configuration ] )
+        .then(function(access) {
+            return access.createMediaKeys();
+        }).then(function(mediaKeys) {
+            config.video.setMediaKeys(mediaKeys);
+            config.video.addEventListener('timeupdate', onTimeupdate, true);
+            consoleWrite(event.data.sessionId);
```
Comment (ddorwin) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: Remove debugging logging?

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+            return _mediaKeySession.load( event.data.sessionId );
+        }).then(function( success ) {
+            if ( !success ) throw new DOMException( 'Could not load session' );
+            return testmediasource(config);
+        }).then(function(source) {
+            config.video.src = URL.createObjectURL(source);
+            config.video.play();
+        })
+        .catch(onFailure);
+
```
Comment (ddorwin) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: empty lines

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 24 above)
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Completion of the `remove()` steps might be delayed while the destruction occurs, but there is no concept of "in use" that would block it. I would expect the playback to fail when that key is needed again. (This seems like a good test case.)

File: encrypted-media/resources/clearkey-retrieve-persistent-license.html (hunk 27 above)
Comment (mwatson2) [inline encrypted-media/resources/clearkey-retrieve-persistent-license.html]: I noted this in the main script file, but not here in the resources file. Thanks for the reminder.

File: encrypted-media/scripts/playback-persistent-license-events.js (hunk 24 above)
Comment (mwatson2) [inline encrypted-media/scripts/playback-persistent-license-events.js]: Hmm, yes, don't know where I got that from about `remove()`. So, here we do not really need to clean up the video element.

**Comment by ddorwin** [comment]
Looks good overall modulo the open comments above. I did not do a detailed review of all the logic.

File: encrypted-media/scripts/playback-persistent-license-events.js [hunk 32]
```diff
@@ -122,8 +122,6 @@ function runTest(config, testname) {
             if ( _video.currentTime > ( config.duration || 5 ) && !_receivedTimeupdateEvent ) {
```
Comment (ddorwin) [inline encrypted-media/scripts/playback-persistent-license-events.js]: I think you changed some of the other tests to 2 already.

**Comment by ddorwin** [comment]
LG.

We might want to address https://github.com/w3c/web-platform-tests/pull/3424#discussion_r75397878 and https://github.com/w3c/web-platform-tests/pull/3424#discussion_r75397906 in a future PR.
